# Exercise 2.2 - Complex Machine Learning Models and Keras

This script contains the following:

1. Importing Libraries and Data
2. Data Preparation (Step 3 in the exercise)
3. Reshaping the Data (Step 4 in the exercise)
4. Split the Data (Step 5 in the exercise)
5. Keras Layered Model - CNN or RNN Model? (Step 2/6 in the exercise)
6. Compiling and Running the Model
7. Confusion Matrix
8. Notes

## 1. Importing Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import os
import tensorflow as tf
from keras.models import Sequential
from tensorflow.keras.layers import Conv1D, Dense, Dropout, Flatten, MaxPooling1D

import matplotlib.pyplot as plt
import seaborn as sns
import operator

In [2]:
path = r'C:\Users\kyles\CareerFoundary\Machine Learning\Achievement 2\02 Data'

In [3]:
# Import unscaled weather data
df_weather = pd.read_csv(os.path.join(path, 'ClimateWins_Dataset_original.csv'))

# Check
df_weather

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0      19600101      1                  7               2.1            0.85   
1      19600102      1                  6               2.1            0.84   
2      19600103      1                  8               2.1            0.90   
3      19600104      1                  3               2.1            0.92   
4      19600105      1                  6               2.1            0.95   
...         ...    ...                ...               ...             ...   
22945  20221027     10                  1               2.1            0.79   
22946  20221028     10                  6               2.1            0.77   
22947  20221029     10                  4               2.1            0.76   
22948  20221030     10                  5               2.1            0.80   
22949  20221031     10                  5               2.1            0.84   

       BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0              1.0180                    0.32                 0.09   
1              1.0180                    0.36                 1.05   
2              1.0180                    0.18                 0.30   
3              1.0180                    0.58                 0.00   
4              1.0180                    0.65                 0.14   
...               ...                     ...                  ...   
22945          1.0248                    1.34                 0.22   
22946          1.0244                    1.34                 0.22   
22947          1.0227                    1.34                 0.22   
22948          1.0212                    1.34                 0.22   
22949          1.0193                    1.34                 0.22   

       BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0                     0             0.7  ...                     5   
1                     0             1.1  ...                     7   
2                     0             0.0  ...                     7   
3                     0             4.1  ...                     7   
4                     0             5.4  ...                     3   
...                 ...             ...  ...                   ...   
22945                 0             7.7  ...                     5   
22946                 0             5.4  ...                     5   
22947                 0             6.1  ...                     5   
22948                 0             5.8  ...                     5   
22949                 0             3.2  ...                     5   

       VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0                   0.88             1.0003                       0.45   
1                   0.91             1.0007                       0.25   
2                   0.91             1.0096                       0.17   
3                   0.86             1.0184                       0.13   
4                   0.80             1.0328                       0.46   
...                  ...                ...                        ...   
22945               0.82             1.0142                       1.13   
22946               0.82             1.0142                       1.13   
22947               0.82             1.0142                       1.13   
22948               0.82             1.0142                       1.13   
22949               0.82             1.0142                       1.13   

       VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0                        0.34                    0                4.7   
1                        0.84                    0                0.7   
2                        0.08                    0                0.1   
3                        0.98                    0                0.0   
4                        0.00                    0                5.7   
...                       ...                  ...   

In [4]:
# Import 'pleasant weather' data
df_pleasant = pd.read_csv(os.path.join(path, 'Pleasant_Weather_Prediction_Answers.csv'))

# Check
df_pleasant

DATE  BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0      19600101                       0                          0   
1      19600102                       0                          0   
2      19600103                       0                          0   
3      19600104                       0                          0   
4      19600105                       0                          0   
...         ...                     ...                        ...   
22945  20221027                       0                          0   
22946  20221028                       0                          0   
22947  20221029                       0                          0   
22948  20221030                       0                          0   
22949  20221031                       0                          0   

       BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   
...                          ...                      ...   
22945                          0                        0   
22946                          0                        0   
22947                          0                        0   
22948                          0                        0   
22949                          0                        0   

       DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                                0                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   
...                            ...                        ...   
22945                            0                          0   
22946                            0                          0   
22947                            0                          0   
22948                            0                          0   
22949                            0                          0   

       KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                            0                           0   
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   
...                        ...                         ...   
22945                        0                           0   
22946                        0                           0   
22947                        0                           0   
22948                        0                           0   
22949                        0                           0   

       MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                                0                        0   
1                                0                        0   
2                                0                        0   
3                                0                        0   
4                                0                        0   
...                            ...                      ...   
22945                            0                        0   
22946                            0                        0   
22947                            0                        0   
22948                            0                        0   
22949                            0                        0   

       MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                              0                      0   
1                              0    

## 2. Data Preparation

The following code ensures the data is correctly structured for the deep learning model. Preparation steps include:

- Drop Unnecessary Columns (DATE & MONTH)
- Drop Unshared Weather Stations
- Drop Observations with Missing Data
- Ensure all Observations Present for All Stations
- Check Data Shapes
- Export Cleaned Data

### 2.1 - Drop Unnecessary Columns

In [5]:
# Drop DATE and MONTH from df_weather
df_weather.drop(columns={'DATE', 'MONTH'}, inplace = True)

# Check
df_weather

BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  BASEL_pressure  \
0                      7               2.1            0.85          1.0180   
1                      6               2.1            0.84          1.0180   
2                      8               2.1            0.90          1.0180   
3                      3               2.1            0.92          1.0180   
4                      6               2.1            0.95          1.0180   
...                  ...               ...             ...             ...   
22945                  1               2.1            0.79          1.0248   
22946                  6               2.1            0.77          1.0244   
22947                  4               2.1            0.76          1.0227   
22948                  5               2.1            0.80          1.0212   
22949                  5               2.1            0.84          1.0193   

       BASEL_global_radiation  BASEL_precipitation  BASEL_snow_depth  \
0                        0.32                 0.09                 0   
1                        0.36                 1.05                 0   
2                        0.18                 0.30                 0   
3                        0.58                 0.00                 0   
4                        0.65                 0.14                 0   
...                       ...                  ...               ...   
22945                    1.34                 0.22                 0   
22946                    1.34                 0.22                 0   
22947                    1.34                 0.22                 0   
22948                    1.34                 0.22                 0   
22949                    1.34                 0.22                 0   

       BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  ...  \
0                 0.7              6.5             0.8  ...   
1                 1.1              6.1             3.3  ...   
2                 0.0              8.5             5.1  ...   
3                 4.1              6.3             3.8  ...   
4                 5.4              3.0            -0.7  ...   
...               ...              ...             ...  ...   
22945             7.7             15.9            11.4  ...   
22946             5.4             16.7            14.3  ...   
22947             6.1             16.7            13.1  ...   
22948             5.8             15.4            11.6  ...   
22949             3.2             13.5             9.9  ...   

       VALENTIA_cloud_cover  VALENTIA_humidity  VALENTIA_pressure  \
0                         5               0.88             1.0003   
1                         7               0.91             1.0007   
2                         7               0.91             1.0096   
3                         7               0.86             1.0184   
4                         3               0.80             1.0328   
...                     ...                ...                ...   
22945                     5               0.82             1.0142   
22946                     5               0.82             1.0142   
22947                     5               0.82             1.0142   
22948                     5               0.82             1.0142   
22949                     5               0.82             1.0142   

       VALENTIA_global_radiation  VALENTIA_precipitation  VALENTIA_snow_depth  \
0                           0.45                    0.34                    0   
1                           0.25                    0.84                    0   
2                           0.17                    0.08                    0   
3                           0.13                    0.98                    0   
4                           0.46                    0.00                    0   
...                          ...                     ...                  ...   
22945                       1.13                    

In [6]:
# drop DATE from df_pleasant
df_pleasant.drop(columns={'DATE'}, inplace = True)

# Check
df_pleasant

BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0                           0                          0   
1                           0                          0   
2                           0                          0   
3                           0                          0   
4                           0                          0   
...                       ...                        ...   
22945                       0                          0   
22946                       0                          0   
22947                       0                          0   
22948                       0                          0   
22949                       0                          0   

       BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   
...                          ...                      ...   
22945                          0                        0   
22946                          0                        0   
22947                          0                        0   
22948                          0                        0   
22949                          0                        0   

       DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                                0                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   
...                            ...                        ...   
22945                            0                          0   
22946                            0                          0   
22947                            0                          0   
22948                            0                          0   
22949                            0                          0   

       KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                            0                           0   
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   
...                        ...                         ...   
22945                        0                           0   
22946                        0                           0   
22947                        0                           0   
22948                        0                           0   
22949                        0                           0   

       MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                                0                        0   
1                                0                        0   
2                                0                        0   
3                                0                        0   
4                                0                        0   
...                            ...                      ...   
22945                            0                        0   
22946                            0                        0   
22947                            0                        0   
22948                            0                        0   
22949                            0                        0   

       MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                              0                      0   
1                              0                      0   
2                              0                      0   
3                              0  

### 2.2 - Drop Unshared Weather Stations

In [7]:
# Create lists for weather stations for both datasets using list comprehensions
stations_in_weather = [col.split('_')[0] for col in df_weather.columns]
stations_in_pleasant = [col.split('_')[0] for col in df_pleasant.columns]

# Find stations in weather list but not in pleasant list
missing_stations = set(stations_in_weather) - set(stations_in_pleasant)

# Print missing stations
print(f'Stations in weather that are not in pleasant: {missing_stations}')

Stations in weather that are not in pleasant: {'TOURS', 'GDANSK', 'ROMA'}


In [8]:
# List all columns for the above weather stations
columns_to_remove = [col for col in df_weather.columns if col.split('_')[0] in missing_stations] # This has been checked

In [9]:
# Remove these columns
df_weather.drop(columns = columns_to_remove, inplace = True)
df_weather

BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  BASEL_pressure  \
0                      7               2.1            0.85          1.0180   
1                      6               2.1            0.84          1.0180   
2                      8               2.1            0.90          1.0180   
3                      3               2.1            0.92          1.0180   
4                      6               2.1            0.95          1.0180   
...                  ...               ...             ...             ...   
22945                  1               2.1            0.79          1.0248   
22946                  6               2.1            0.77          1.0244   
22947                  4               2.1            0.76          1.0227   
22948                  5               2.1            0.80          1.0212   
22949                  5               2.1            0.84          1.0193   

       BASEL_global_radiation  BASEL_precipitation  BASEL_snow_depth  \
0                        0.32                 0.09                 0   
1                        0.36                 1.05                 0   
2                        0.18                 0.30                 0   
3                        0.58                 0.00                 0   
4                        0.65                 0.14                 0   
...                       ...                  ...               ...   
22945                    1.34                 0.22                 0   
22946                    1.34                 0.22                 0   
22947                    1.34                 0.22                 0   
22948                    1.34                 0.22                 0   
22949                    1.34                 0.22                 0   

       BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  ...  \
0                 0.7              6.5             0.8  ...   
1                 1.1              6.1             3.3  ...   
2                 0.0              8.5             5.1  ...   
3                 4.1              6.3             3.8  ...   
4                 5.4              3.0            -0.7  ...   
...               ...              ...             ...  ...   
22945             7.7             15.9            11.4  ...   
22946             5.4             16.7            14.3  ...   
22947             6.1             16.7            13.1  ...   
22948             5.8             15.4            11.6  ...   
22949             3.2             13.5             9.9  ...   

       VALENTIA_cloud_cover  VALENTIA_humidity  VALENTIA_pressure  \
0                         5               0.88             1.0003   
1                         7               0.91             1.0007   
2                         7               0.91             1.0096   
3                         7               0.86             1.0184   
4                         3               0.80             1.0328   
...                     ...                ...                ...   
22945                     5               0.82             1.0142   
22946                     5               0.82             1.0142   
22947                     5               0.82             1.0142   
22948                     5               0.82             1.0142   
22949                     5               0.82             1.0142   

       VALENTIA_global_radiation  VALENTIA_precipitation  VALENTIA_snow_depth  \
0                           0.45                    0.34                    0   
1                           0.25                    0.84                    0   
2                           0.17                    0.08                    0   
3                           0.13                    0.98                    0   
4                           0.46                    0.00                    0   
...                          ...                     ...                  ...   
22945                       1.13                    

### 2.3 - Drop Observations with Missing Data

In [10]:
missing_values = df_weather.isnull().sum()
missing_columns = missing_values[missing_values > 0]

if missing_columns.empty:
    print("There are no missing values in any column.")
else:
    print("Columns with missing values and their counts:")
    print(missing_columns)

There are no missing values in any column.


Perhaps the instruction in the exercise meant that some of the weather observations are missing entirely for multiple stations. Each weather metric should be checked to see how many times it occurs across the full data set.

In [11]:
# Create list of all weather observation types
weather_observations = list(set([col.split('_', 1)[1] for col in df_weather.columns])) # Use col.split('_', 1) to limit the number of splits
weather_observations = sorted(weather_observations)
weather_observations

['cloud_cover',
 'global_radiation',
 'humidity',
 'precipitation',
 'pressure',
 'snow_depth',
 'sunshine',
 'temp_max',
 'temp_mean',
 'temp_min',
 'wind_speed']

In [12]:
# Count number of times each weather metric occurs in the title of all columns
print(f'Number of Stations covered by each weather observation:')

for ob in weather_observations:
    columns = [col for col in df_weather if col.endswith(ob)]
    count = len(columns)
    print(f'{ob}: {count} stations')

Number of Stations covered by each weather observation:
cloud_cover: 14 stations
global_radiation: 15 stations
humidity: 14 stations
precipitation: 15 stations
pressure: 14 stations
snow_depth: 6 stations
sunshine: 15 stations
temp_max: 15 stations
temp_mean: 15 stations
temp_min: 15 stations
wind_speed: 9 stations


The 2 columns which should be dropped are **snow_depth** and **wind_speed**

In [13]:
# List columns with these observations
columns_to_drop = [col for col in df_weather if col.endswith('snow_depth') or col.endswith('wind_speed')]
print(columns_to_drop) # Ensure correct columns have been assigned
print(len(columns_to_drop)) # Ensure 15 columns are included

['BASEL_wind_speed', 'BASEL_snow_depth', 'DEBILT_wind_speed', 'DUSSELDORF_wind_speed', 'DUSSELDORF_snow_depth', 'HEATHROW_snow_depth', 'KASSEL_wind_speed', 'LJUBLJANA_wind_speed', 'MAASTRICHT_wind_speed', 'MADRID_wind_speed', 'MUNCHENB_snow_depth', 'OSLO_wind_speed', 'OSLO_snow_depth', 'SONNBLICK_wind_speed', 'VALENTIA_snow_depth']
15


In [14]:
# Drop columns
df_weather.drop(columns = columns_to_drop, inplace = True)
df_weather.shape # Should be down to 132 columns

(22950, 132)

### 2.4 - Ensure all Observations Present for All Stations

**pressure**, **humidity**, and **cloud_cover** are each missing from only 1 out of 15 weather stations. This data should be supplemented (copied) from a nearby weather station. Start with finding which stations are missing each observation.

In [15]:
# Create a list containing the above weather observations
missing_observations = ['cloud_cover', 'humidity', 'pressure']

In [16]:
# Make a list of all stations 
all_stations = sorted(list(set([col.split('_')[0] for col in df_weather.columns]))) 
# Check
print(len(all_stations))

15


In [17]:
# Create dictionary for storing station and missing observation
missing_stations_by_observation = {}

for ob in missing_observations:
    # Start with selecting stations where the observation is present
    stations_covered = [col.split('_')[0] for col in df_weather.columns if col.endswith(ob)]

    # Identify missing station by subtracting covered stations from all stations 
    missing_station = set(all_stations) - set(stations_covered)

    # Store station and missing observation in dictionary
    missing_stations_by_observation[ob] = missing_station

# Print stations along with their missing observation
for ob, missing_station in missing_stations_by_observation.items():
    print(f'Station missing {ob}: {", ".join(missing_station)}')

Station missing cloud_cover: KASSEL
Station missing humidity: STOCKHOLM
Station missing pressure: MUNCHENB


In [18]:
# Check head of df_weather to assist in determining placement of new columns
df_weather.head()

BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  BASEL_global_radiation  \
0                  7            0.85           1.018                    0.32   
1                  6            0.84           1.018                    0.36   
2                  8            0.90           1.018                    0.18   
3                  3            0.92           1.018                    0.58   
4                  6            0.95           1.018                    0.65   

   BASEL_precipitation  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                 0.09             0.7              6.5             0.8   
1                 1.05             1.1              6.1             3.3   
2                 0.30             0.0              8.5             5.1   
3                 0.00             4.1              6.3             3.8   
4                 0.14             5.4              3.0            -0.7   

   BASEL_temp_max  BELGRADE_cloud_cover  ...  STOCKHOLM_temp_max  \
0            10.9                     1  ...                 4.9   
1            10.1                     6  ...                 5.0   
2             9.9                     6  ...                 4.1   
3            10.6                     8  ...                 2.3   
4             6.0                     8  ...                 4.3   

   VALENTIA_cloud_cover  VALENTIA_humidity  VALENTIA_pressure  \
0                     5               0.88             1.0003   
1                     7               0.91             1.0007   
2                     7               0.91             1.0096   
3                     7               0.86             1.0184   
4                     3               0.80             1.0328   

   VALENTIA_global_radiation  VALENTIA_precipitation  VALENTIA_sunshine  \
0                       0.45                    0.34                4.7   
1                       0.25                    0.84                0.7   
2                       0.17                    0.08                0.1   
3                       0.13                    0.98                0.0   
4                       0.46                    0.00                5.7   

   VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
0                 8.5                6.0               10.9  
1                 8.9                5.6               12.1  
2                10.5                8.1               12.9  
3                 7.4                7.3               10.6  
4                 5.7                3.0                8.4  

[5 rows x 132 columns]

In [19]:
all_stations

['BASEL',
 'BELGRADE',
 'BUDAPEST',
 'DEBILT',
 'DUSSELDORF',
 'HEATHROW',
 'KASSEL',
 'LJUBLJANA',
 'MAASTRICHT',
 'MADRID',
 'MUNCHENB',
 'OSLO',
 'SONNBLICK',
 'STOCKHOLM',
 'VALENTIA']

In [20]:
# Check position of HEATHROW_temp_max to find where we need to place new KASSEL_cloud_cover column
print(f'KASSEL_cloud_cover column position: {df_weather.columns.get_loc("HEATHROW_temp_max") + 1}') # +1 to position column to the right

# Check position of MUNCHENB_humidity to find where we need to place new MUNCHENB_pressure column
print(f'MUNCHENB_pressure column position: {df_weather.columns.get_loc("MUNCHENB_humidity") + 2}') # +2 (consider Kassel's new column)

# Check position of STOCKHOLM_cloud_cover to find where we need to place new STOCKHOLM_humidity column
print(f'STOCKHOLM_humidity column position: {df_weather.columns.get_loc("STOCKHOLM_cloud_cover") + 3}') # +3 (as above)

KASSEL_cloud_cover column position: 54
MUNCHENB_pressure column position: 92
STOCKHOLM_humidity column position: 118


In [21]:
# Finally, insert new columns into df_weather
# Copy data from nearby weather stations: 
# Kassel with Ljubljana, Munchenb with Sonnblick, Stockholm with Oslo

df_weather.insert(54, 'KASSEL_cloud_cover', df_weather['LJUBLJANA_cloud_cover'])
df_weather.insert(92, 'MUNCHENB_pressure', df_weather['SONNBLICK_pressure'])
df_weather.insert(118, 'STOCKHOLM_humidity', df_weather['OSLO_humidity'])

In [22]:
# Check
df_weather.columns.tolist()

['BASEL_cloud_cover',
 'BASEL_humidity',
 'BASEL_pressure',
 'BASEL_global_radiation',
 'BASEL_precipitation',
 'BASEL_sunshine',
 'BASEL_temp_mean',
 'BASEL_temp_min',
 'BASEL_temp_max',
 'BELGRADE_cloud_cover',
 'BELGRADE_humidity',
 'BELGRADE_pressure',
 'BELGRADE_global_radiation',
 'BELGRADE_precipitation',
 'BELGRADE_sunshine',
 'BELGRADE_temp_mean',
 'BELGRADE_temp_min',
 'BELGRADE_temp_max',
 'BUDAPEST_cloud_cover',
 'BUDAPEST_humidity',
 'BUDAPEST_pressure',
 'BUDAPEST_global_radiation',
 'BUDAPEST_precipitation',
 'BUDAPEST_sunshine',
 'BUDAPEST_temp_mean',
 'BUDAPEST_temp_min',
 'BUDAPEST_temp_max',
 'DEBILT_cloud_cover',
 'DEBILT_humidity',
 'DEBILT_pressure',
 'DEBILT_global_radiation',
 'DEBILT_precipitation',
 'DEBILT_sunshine',
 'DEBILT_temp_mean',
 'DEBILT_temp_min',
 'DEBILT_temp_max',
 'DUSSELDORF_cloud_cover',
 'DUSSELDORF_humidity',
 'DUSSELDORF_pressure',
 'DUSSELDORF_global_radiation',
 'DUSSELDORF_precipitation',
 'DUSSELDORF_sunshine',
 'DUSSELDORF_temp_mean',


### 2.5 - Check Data Shapes

In [23]:
df_weather.shape

(22950, 135)

In [24]:
df_pleasant.shape

(22950, 15)

Both data sets have the correct shape

### 2.6 - Export Cleaned Data

In [25]:
df_weather.to_csv(os.path.join(path, 'Cleaned_Weather_Data.csv'))

## 3. Reshaping the Data

Split observations (df_weather) into **15 groups** of **9 types of observations**, and labels (df_pleasant) should also be in **15 groups**. That is:

- Shape of X = (22950, 15, 9) and y = (22950, 15)

In [26]:
X = df_weather
y = df_pleasant

In [27]:
# Turn X and y into arrays
X = np.array(X)
y = np.array(y)

In [28]:
# Reshape X (don't need to reshape y)
X = X.reshape(-1, 15, 9)

# Check
X.shape

(22950, 15, 9)

## 4. Split the Data

In [29]:
# Import train_test_split
from sklearn.model_selection import train_test_split

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42) # Can also control size with train_size before random_state

In [30]:
# Check sizes
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(17212, 15, 9) (17212, 15)
(5738, 15, 9) (5738, 15)


## 5. Keras Layered Model - Convolution Neural Network (CNN)

In [315]:
# Define hyperparamters at the top for easy adjustments
epochs = 30
batch_size = 16
n_hidden = 128

timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = len(y_train[0])

model = Sequential()
model.add(Conv1D(n_hidden, kernel_size=2, activation='relu', input_shape=(timesteps, input_dim)))
model.add(Dense(16, activation='relu'))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax')) # options include: softmax, sigmoid, tanh, or relu

C:\Users\kyles\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [316]:
model.summary()

Model: "sequential_41"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d_41 (Conv1D)              │ (None, 14, 128)        │         2,432 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_82 (Dense)                │ (None, 14, 16)         │         2,064 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_41 (MaxPooling1D) │ (None, 7, 16)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_41 (Flatten)            │ (None, 112)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_83 (Dense)                │ (None, 15)             │         1,695 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 6,191 (24.18 KB)

 Trainable params: 6,191 (24.18 KB)

 Non-trainable params: 0 (0.00 B)

## 6. Compiling and Running the Model

In [318]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [319]:
model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)

Epoch 1/30
1076/1076 - 3s - 3ms/step - accuracy: 0.1199 - loss: 20605.7871
Epoch 2/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1333 - loss: 214349.6094
Epoch 3/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1323 - loss: 702596.3125
Epoch 4/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1310 - loss: 1530995.3750
Epoch 5/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1322 - loss: 2741419.0000
Epoch 6/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1326 - loss: 4432945.0000
Epoch 7/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1315 - loss: 6475533.5000
Epoch 8/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1300 - loss: 9153262.0000
Epoch 9/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1299 - loss: 11976268.0000
Epoch 10/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1321 - loss: 15575165.0000
Epoch 11/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1307 - loss: 19517938.0000
Epoch 12/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1307 - loss: 24527042.0000
Epoch 13/30
1076/1076 - 1s - 1ms/step - accuracy: 0.1229 - loss: 29626266

## 7. Confusion Matrix

In [320]:
# Set up 'label key' for confusion matrix (similar to activities in HAR Data)
stations = {
    0: 'BASEL', 
    1: 'BELGRADE', 
    2: 'BUDAPEST', 
    3: 'DEBILT', 
    4: 'DUSSELDORF', 
    5: 'HEATHROW', 
    6: 'KASSEL', 
    7: 'LJUBLJANA', 
    8: 'MAASTRICHT', 
    9: 'MADRID', 
    10: 'MUNCHENB', 
    11: 'OSLO', 
    12: 'SONNBLICK', 
    13: 'STOCKHOLM', 
    14: 'VALENTIA'
}

In [321]:
def confusion_matrix(Y_true, Y_pred):
    Y_true = pd.Series([stations[y] for y in np.argmax(Y_true, axis=1)])
    Y_pred = pd.Series([stations[y] for y in np.argmax(Y_pred, axis=1)])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

# Evaluate
print(confusion_matrix(y_test, model.predict(X_test)))

180/180 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step  
Pred        BASEL  BELGRADE  BUDAPEST  DEBILT  DUSSELDORF  HEATHROW  KASSEL  \
True                                                                          
BASEL         322       147       348     667          63        54       9   
BELGRADE       44        28       190      84           9         5       0   
BUDAPEST        2         4        22      30           2         2       0   
DEBILT          1         0         9      20           0         0       0   
DUSSELDORF      0         0         2      11           0         1       0   
HEATHROW        0         0         1      24           0         1       0   
KASSEL          0         0         3       2           0         0       0   
LJUBLJANA       1         0        12       4           0         1       0   
MAASTRICHT      0         0         1       0           1         0       0   
MADRID         11         1        18      91           1        21       0   
MUNCHENB 

## 8. Notes:

- What happens to loss and accuracy for different combinations of hyperparameters?
- Does the model converge or does the loss grow exponentially? (list the activation type used)
- How accurate is the model at recognising the stations?


### 8.1 - Changing Activation Types

**Attempt 1:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***softmax***

**Results:** loss increased exponentially. Accuracy hovered around 16% to 11%. "Recognised" 9/15 stations poorly

**Attempt 2:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***sigmoid***

**Results:** loss increased exponentially. Accuracy hovered around 11% to 13%. "Recognised" 1 station only (Basel)

**Attempt 3:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***tanh***

**Results:** loss changed only from epoch 1 to epoch 4 (decreased slightly, stabalised at 22.9). Accuracy increased but stayed at 25.5% after epoch 4. "Recognised" 5 stations only.

**Attempt 4:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***relu***

**Results:** loss became "nan" at epoch 2. Accuracy increased to 64.4% at epoch 3. This appears to be some kind of error

***Thoughts:*** perhaps the number of layers is too low right now for the model to create enough meaningful connections within the data and thus the activation types are just creating different (perhaps non-meaningful) errors because of that.

### 8.2 - Increasing Hidden Layers with Sigmoid

**Attempt 5:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***8***, 

activation type = sigmoid

**Results:** loss increased exponentially. Accuracy ≈ 13% to 14%. "Recognised" 2 stations only.

**Attempt 6:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***16***, 

activation type = sigmoid

**Results:** loss increased exponentially. Accuracy ≈ 11% to 12%. "Recognised" 1 station only.

**Attempt 7:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***32***, 

activation type = sigmoid

**Results:** loss increased exponentially. Accuracy ≈ 13% to 14%. "Recognised" 1 stations only.

***Thoughts:*** Attempts to further increase higgen layers led to similar results. Changing back to softmax at a higher starting number of hidden layers

### 8.3 - Increasing Hidden Layers with Softmax

***Summary:*** Basically, the loss function continued to grow exponentially despite the number of hidden layers. Although, 15 stations recogised after having 8 layers. Will now try tanh.

### 8.4 - Increasing Hidden Layers with Tanh

**Attempt 12:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***16***, 

activation type = tanh

**Results:** loss changed only from epoch 1 to epoch 14 (decreased slightly, stabalised at 22.8). Accuracy increased but stayed at 24.8% after epoch 14. "Recognised" 6 stations only.

***Summary of next attemtpts:*** Increasing the number of hidden layers made the model less accurate, only ever "recognising" 8 stations out of 15. loss never increased exponentially but stopped changing early into the epochs

### 8.5 - Increasing Hidden Layers with Relu

***Summary of next attempts:*** Increasing hidden layers also didn't help. Accuracy stays at 64.4%, loss goes to "nan", and matrix only recognises 1 station.

### 8.6 - Final Attempts with Softmax

Softmax seems to show the most promise for the activation type as it allows the model to seemingly identify the highest number of stations. I will experiment with the other hyperparameters using softmax before trialling out the RNN Model

***8 layers:*** 15 stations, Accuracy ≈ 10% to 14%, loss --> Exponential

***16 layers:*** 15 stations, Accuracy ≈ 10% to 14%, loss --> Exponential

***32 layers:*** 10 stations, Accuracy ≈ 10% to 14%, loss --> Exponential

***64 layers:*** 13 stations, Accuracy ≈ 12% to 14%, loss --> Exponential

***128 layers:*** 15 stations, Accuracy ≈ 10% to 13%, loss --> Exponential

Final note: After rerunning the same hyperparameters again for the purposes of obtaining screenshots, I have come to realise the model is somewhat random and may present slightly different results each time. These are usually minor (like 1% change to final accuracy, but can also be more significant, like failing to recognise multiple stations that were previously recognised. Perhaps CNN isn't so good for this